src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -398,7 +487,93 @@ internal void Run<TState>(string input, int startat, ref TState state, MatchCall
+                        {
+                            // If the callback returns false, we're done.
+                            // Drop reference to text to avoid keeping it alive in a cache.
+                            runner.runtext = null!;
+                            if (reuseMatchObject)
+                            {
+                                // We're reusing the single match instance, so clear out its text as well.
+                                // We don't do this if we're not reusing instances, as in that case we're
+                                // dropping the whole reference to the match, and we no longer own the instance
+                                // having handed it out to the callback.
+                                m.Text = null!;
+                            }
+                            return;
+                        }
+
+                        // Now that we've matched successfully, update the starting position to reflect
+                        // the current position, just as Match.NextMatch() would pass in _textpos as textstart.
+                        runtextpos = startat = runner.runtextpos;
+
+
+                        // Reset state for another iteration.
+                        runner.runtrackpos = runner.runtrack!.Length;
+                        runner.runstackpos = runner.runstack!.Length;
+                        runner.runcrawlpos = runner.runcrawl!.Length;
+
+                        if (m.Length == 0)
+                        {
+                            if (runner.runtextpos == stoppos)
+                            {
+                                // Drop reference to text to avoid keeping it alive in a cache.
+                                runner.runtext = null!;
+                                if (reuseMatchObject)
+                                {
+                                    // See above comment.
+                                    m.Text = null!;
+                                }
+                                return;
+                            }
+
+                            runtextpos += RightToLeft ? -1 : 1;
+                        }
+
+                        // Loop around to perform next match from where we left off.
+                        continue;
+                    }
+                    else
+                    {
+                        // We failed to match at this position.  If we're at the stopping point, we're done.
+                        if (runner.runtextpos == stoppos)
+                        {
+                            runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+                            if (runner.runmatch != null)
+                            {
+                                runner.runmatch.Text = null!;
+                            }
+                            return;
+                        }
+                    }
+                }
```

Comment:
Can we implement this method in terms of the other Run overload?  It'd be nice to avoid all this complicated duplication.  I realize it'll come at a small cost, but hopefully it'd be fairly minor.

Reply:
Sure, I can do that.